# COMPARATIVE STUDY OF DIFFERENT MACHINE LEARNING MODELS FOR SALES PREDICTION AND FRAUD DETECTION.

# Introduction

In recent years, the rise of the Internet of things (IoT) as an emerging technology has been unbelievable, more companies are moving towards the adoption of these technologies and many IoT sensors are being deployed to share information in real-time which leads to the generation of a huge amount of data. This data when used correctly, will be very helpful to the company to discover hidden patterns for better decision making in the future. For example, with the DataCo company, dataset customer segmentation analysis was performed in this project which helps the company to better understand its customers and target them to increase customer responsiveness and the company's revenue. With a lot of options available to analyze data, it is very difficult to decide which method and machine learning model to use since the performance of the model vary on the parameters available in the data. 

With the growth of machine learning, there have been numerous comparison studies that compare the performance of neural networks with traditional linear techniques for forecasting. For example, author Carbonneau et al. (2007) in his research work compared various traditional forecasting time-series like moving average, linear regression with recurrent neural networks and support vector machines and concluded that recurrent neural networks performed best. Hill et al. (1996) have also considered the M-competition data and have compared between neural networks and traditional methods. Vakili et al. (2020) evaluated the performance of 11 popular machine and deep learning algorithms for classification task using six IoT-related datasets and concluded that Random Forests performed better than other machine learning models, while among deep learning models, ANN and CNN achieved more interesting results. Some other authors like Ahmed et al. (2010) did study comparing different regression models and concluded that the MLP model and Gaussian process models are the best two models for regression type data. But no study that compared both Classification type ML models and Regression type ML models against the Neural Network models with the same dataset was found.

This project aims to compare 9 popular machine learning classifiers and 7 regressors type machine learning models and measure their performance against neural network models to find out which machine learning model performs better. Since the dataset used is related to supply chain important parameters are identified and the machine learning models are trained with the dataset for detection of fraud transactions, late delivery of orders, sales revenue and quantity of products which customer orders. The machine learning classifiers used in this project are Logistic Regression,Linear Discriminant Analysis, Gaussian Naive Bayes, Support Vector Machines, k - Nearest Neighbors, Random Forest classification, Extra Trees classification, Extreme Gradient Boosting, Decision Tree classification for fraud detection and to predict late delivery on the basis accuracy, recall score and F1 score. The regression models used are Lasso, Ridge, Light Gradient boosting, Random Forest regression, Extreme Gradient Boosting regression, Decision Tree Regression, and Linear Regression to predict sales and quantity of the products required which are compared with mean absolute error (MAE) and root mean square error (RMSE).

# Data Collection

The dataset used in this project is maintained transparently with the Creative Commons 4.0 license by Fabian Constante, Fernando Silva, and António Pereira through the Mendeley data repository. The dataset consists of roughly 180k transactions from supply chains used by the company DataCo Global for 3 years. The dataset can be downloaded from:

https://data.mendeley.com/datasets/8gx2fvg2k6/5

### Importing all required libraries

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import xgboost as xgb
import lightgbm as lgb
import datetime as dt
import calendar,warnings,itertools,matplotlib,keras,shutil
import tensorflow as tf
import statsmodels.api as sm
from datetime import datetime
from sklearn.model_selection import train_test_split,cross_val_score, cross_val_predict
from sklearn import svm,metrics,tree,preprocessing,linear_model
from sklearn.preprocessing import MinMaxScaler,StandardScaler
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import Ridge,LinearRegression,LogisticRegression,ElasticNet, Lasso
from sklearn.ensemble import RandomForestRegressor,RandomForestClassifier, GradientBoostingRegressor,BaggingClassifier,ExtraTreesClassifier
from sklearn.metrics import accuracy_score,mean_squared_error,recall_score,confusion_matrix,f1_score,roc_curve, auc
from sklearn.datasets import load_iris,make_regression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.kernel_ridge import KernelRidge
from keras import Sequential
from keras.layers import Dense
from IPython.core import display as ICD
#from tensorflow.estimator import inputs

#Hiding the warnings
warnings.filterwarnings('ignore') 


In [5]:
#Importing Dataset using pandas
file_path = r"C:\Users\User\Downloads\DataCoSupplyChainDataset.csv"

# Load the data
dataset = pd.read_csv(file_path, header=0, encoding='unicode_escape')

# Verify the load
dataset.head(5)

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,...,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


# Data Cleaning

In [6]:
dataset.shape

(180519, 53)

The total data set consists of 180519 records and 53 columns

In [7]:
dataset.apply(lambda x: sum(x.isnull())) #Checking missing values

Type                                  0
Days for shipping (real)              0
Days for shipment (scheduled)         0
Benefit per order                     0
Sales per customer                    0
Delivery Status                       0
Late_delivery_risk                    0
Category Id                           0
Category Name                         0
Customer City                         0
Customer Country                      0
Customer Email                        0
Customer Fname                        0
Customer Id                           0
Customer Lname                        8
Customer Password                     0
Customer Segment                      0
Customer State                        0
Customer Street                       0
Customer Zipcode                      3
Department Id                         0
Department Name                       0
Latitude                              0
Longitude                             0
Market                                0


The data consists of some missing values from Customer Lname, Product Description, Order Zipcode and, Customer Zipcode which should be removed or replaced before proceeding with the analysis. And also, since there is a chance different customers might have the same first name or same last name a new column with ‘customer full name’ is created to avoid any ambiguities.

In [8]:
# Adding first name and last name together to create new column
dataset['Customer Full Name'] = dataset['Customer Fname'].astype(str)+dataset['Customer Lname'].astype(str)

To make it easier for analysis some unimportant columns are dropped

In [9]:
data=dataset.drop(['Customer Email','Product Status','Customer Password','Customer Street','Customer Fname','Customer Lname',
           'Latitude','Longitude','Product Description','Product Image','Order Zipcode','shipping date (DateOrders)'],axis=1)
data.shape

(180519, 42)


There are 3 missing values in Customer Zipcode column. Since the missing values are just zip codes which are not very important these are replaced with zero before proceeding with data analysis.

In [10]:
data['Customer Zipcode']=data['Customer Zipcode'].fillna(0)#Filling NaN columns with zero

# Data Visualisation

<!-- To find important parameters, data correlation is performed.<p>
NOTE: Since this code block does not directly contribute to the comparison of the seected model (Decision Tree) most of it was commented out. -->

In [11]:
# fig, ax = plt.subplots(figsize=(24,12))         # figsize
# sns.heatmap(data.corr(),annot=True,linewidths=.5,fmt='.1g',cmap= 'coolwarm') # Heatmap for correlation matrix

In [12]:
# market = data.groupby('Market') Grouping by market
# region = data.groupby('Order Region')
# plt.figure(1)
#market['Sales per customer'].sum().sort_values(ascending=False).plot.bar(figsize=(12,6), title="Total sales for all markets")
#plt.figure(2)
#region['Sales per customer'].sum().sort_values(ascending=False).plot.bar(figsize=(12,6), title="Total sales for all regions")

In [13]:
#Grouping all categories
#cat = data.groupby('Category Name')
#plt.figure(1)
# Total sum of sales for all categories
#cat['Sales per customer'].sum().sort_values(ascending=False).plot.bar(figsize=(12,6), title="Total sales")
# Mean sales for all categories
#plt.figure(2)
#cat['Sales per customer'].mean().sort_values(ascending=False).plot.bar(figsize=(12,6), title="Average sales")
#plt.figure(3)
# Mean prices for all categories
#cat['Product Price'].mean().sort_values(ascending=False).plot.bar(figsize=(12,6), title="Average price")

In [14]:
#data.plot(x='Product Price', y='Sales per customer',linestyle='dotted',
     #markerfacecolor='blue', markersize=12) 
#plt.title('Product Price vs Sales per customer')#title
#plt.xlabel('Product Price')  # X-axis title
#plt.ylabel('Sales per customer') # Y=axis title
#plt.show()

Hidden Feature Creation

In [46]:

data['order_year']= pd.DatetimeIndex(data['order date (DateOrders)']).year
data['order_month'] = pd.DatetimeIndex(data['order date (DateOrders)']).month
data['order_week_day'] = pd.DatetimeIndex(data['order date (DateOrders)']).day_name()
data['order_hour'] = pd.DatetimeIndex(data['order date (DateOrders)']).hour
data['order_month_year'] = pd.to_datetime(data['order date (DateOrders)']).dt.to_period('M')

In [16]:
#quater= data.groupby('order_month_year')
#quartersales=quater['Sales'].sum().resample('Q').mean().plot(figsize=(15,6))

In [17]:
#plt.figure(figsize=(10,12))
#plt.subplot(4, 2, 1)
#quater= data.groupby('order_year')
#quater['Sales'].mean().plot(figsize=(12,12),title='Average sales for 3 years')
#plt.subplot(4, 2, 2)
#days=data.groupby("order_week_day")
#days['Sales'].mean().plot(figsize=(12,12),title='Average sales per week in days')
#plt.subplot(4, 2, 3)
#hrs=data.groupby("order_hour")
#hrs['Sales'].mean().plot(figsize=(12,12),title='Average sales per day in hrs')
#plt.subplot(4, 2, 4)
#mnth=data.groupby("order_month")
#mnth['Sales'].mean().plot(figsize=(12,12),title='Average sales per year in mnths')
#plt.tight_layout()
#plt.show()

In [18]:
#data['Type'].unique()

In [19]:
#xyz = data.groupby('Type')
#xyz1 = data[(data['Type'] == 'TRANSFER')]
#xyz2= data[(data['Type'] == 'CASH')]
#xyz3= data[(data['Type'] == 'PAYMEN#T')]
#xyz4= data[(data['Type'] == 'DEBIT')]
#count1=xyz1['Order Region'].value_counts()
#count2=xyz2['Order Region'].value_counts()
#count3=xyz3['Order Region'].value_counts()
#count4=xyz4['Order Region'].value_counts()
#names=data['Order Region'].value_counts().keys()
#n_groups=23
#fig,ax = plt.subplots(figsize=(20,8))
#index=np.arange(n_groups)
#bar_width=0.2
#opacity=0.6
#type1=plt.bar(index,count1,bar_width,alpha=opacity,color='b',label='Transfer')
#type2=plt.bar(index+bar_width,count2,bar_width,alpha=opacity,color='r',label='Cash')
#type3=plt.bar(index+bar_width+bar_width,count3,bar_width,alpha=opacity,color='g',label='Payment')
#type4=plt.bar(index+bar_width+bar_width+bar_width,count4,bar_width,alpha=opacity,color='y',label='Debit')
#plt.xlabel('Order Regions')
#plt.ylabel('Number of payments')
#plt.title('Different Type of payments used in all regions')
#plt.legend()
#plt.xticks(index+bar_width,names,rotation=90)
#plt.tight_layout()
#plt.show()


In [20]:
#loss = data[(data['Benefit per order']<0)]
#Plotting top 10 products with most loss
#plt.figure(1)
#loss['Category Name'].value_counts().nlargest(10).plot.bar(figsize=(20,8), title="Products with most loss")
#plt.figure(2)
#loss['Order Region'].value_counts().nlargest(10).plot.bar(figsize=(20,8), title="Regions with most loss")
#Sum of total sales which are lost
#print('Total revenue lost with orders',loss['Benefit per order'].sum())

In [21]:
#Checking type of payment used to conduct fraud other than Transfer
#xyz = data[(data['Type'] != 'TRANSFER')&(data['Order Status'] == 'SUSPECTED_FRAUD')]
#xyz['Order Region'].value_counts()

In [22]:
#high_fraud = data[(data['Order Status'] == 'SUSPECTED_FRAUD') & (data['Type'] == 'TRANSFER')]#separating orders with suspected fraud
#Plotting pie chart with respect to order region
#fraud=high_fraud['Order Region'].value_counts().plot.pie(figsize=(24,12),
 #                                                  startangle=180, explode=(0.1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0),autopct='%.1f',shadow=True,)
#plt.title("Regions with Highest Fraud",size=15,color='r') # Plotting title
#plt.ylabel(" ")
#fraud.axis('equal') 
#plt.show()

In [23]:
#high_fraud1 = data[(data['Order Status'] == 'SUSPECTED_FRAUD')] # 
#high_fraud2 = data[(data['Order Status'] == 'SUSPECTED_FRAUD') &(data['Order Region'] == 'Western Europe')]
#Plotting bar chart for top 10 most suspected fraud department in all regions
#fraud1=high_fraud1['Category Name'].value_counts().nlargest(10).plot.bar(figsize=(20,8), title="Fraud Category",color='orange')
#Plotting bar chart for top 10 most suspected fraud department in Western Europe
#fraud2=high_fraud2['Category Name'].value_counts().nlargest(10).plot.bar(figsize=(20,8), title="Fraud product in Western Europe",color='green')
#plt.legend(["All regions", "Western Europe"])
#plt.title("Top 10 products with highest fraud detections", size=15)
#plt.xlabel("Products", size=13)
#plt.ylim(0,600)
#plt.show()

In [24]:
#Filtering out suspected fruad orders
#cus = data[(data['Order Status'] == 'SUSPECTED_FRAUD')] 
#Top 10 customers with most fraud
#cus['Customer Full Name'].value_counts().nlargest(10).plot.bar(figsize=(20,8), title="Top 10 Highest Fraud Customers")

In [25]:
#Filtering orders of mary smith with suspected fraud
#amount = data[(data['Customer Full Name'] == 'MarySmith')&(data['Order Status'] == 'SUSPECTED_FRAUD')]
#Plotting bar chart for top 10 most suspected fraud customers
#amount['Sales'].sum()

In [26]:
#Filtering columns with late delivery status
#late_delivery = data[(data['Delivery Status'] == 'Late delivery')]
#Top 10 products with most late deliveries
#late_delivery['Category Name'].value_counts().nlargest(10).plot.bar(figsize=(20,8), title="Top 10 products with most late deliveries")


In [27]:
#Filtering orders with late delivery risk
#xyz1 = data[(data['Late_delivery_risk'] == 1)]
#Filtering late delivered orders
#xyz2 = data[(data['Delivery Status'] == 'Late delivery')]
#count1=xyz1['Order Region'].value_counts()
#count2=xyz2['Order Region'].value_counts()
#Index names
#names=data['Order Region'].value_counts().keys()
#n_groups=23
#fig,ax = plt.subplots(figsize=(20,8)) #Figure size
#index=np.arange(n_groups)
#bar_width=0.2
#opacity=0.8
#type1=plt.bar(index,count1,bar_width,alpha=opacity,color='r',label='Risk of late delivery')
#type2=plt.bar(index+bar_width,count2,bar_width,alpha=opacity,color='y',label='Late delivery')
#plt.xlabel('Order Regions')
#plt.ylabel('Number of shipments')
#plt.title('Late delivered products used in all regions')
#plt.legend()
#plt.xticks(index+bar_width,names,rotation=90)
#plt.tight_layout()
#plt.show()


In [28]:
#Filtering late delivery orders with standard class shipping
#xyz1 = data[(data['Delivery Status'] == 'Late delivery') & (data['Shipping Mode'] == 'Standard Class')]
#Filtering late delivery orders with first class shipping
#xyz2 = data[(data['Delivery Status'] == 'Late delivery') & (data['Shipping Mode'] == 'First Class')]
#Filtering late delivery orders with second class shipping
#xyz3 = data[(data['Delivery Status'] == 'Late delivery') & (data['Shipping Mode'] == 'Second Class')]
#Filtering late delivery orders with same day shipping
#xyz4 = data[(data['Delivery Status'] == 'Late delivery') & (data['Shipping Mode'] == 'Same Day')]
#Counting total values
#count1=xyz1['Order Region'].value_counts()
#count2=xyz2['Order Region'].value_counts()
#count3=xyz3['Order Region'].value_counts()
#count4=xyz4['Order Region'].value_counts()
#Index names
#names=data['Order Region'].value_counts().keys()
#n_groups=23
#fig,ax = plt.subplots(figsize=(20,8))
#index=np.arange(n_groups)
#bar_width=0.2
#opacity=0.6
#type1=plt.bar(index,count1,bar_width,alpha=opacity,color='b',label='Standard Class')
#type2=plt.bar(index+bar_width,count2,bar_width,alpha=opacity,color='r',label='First class')
#type3=plt.bar(index+bar_width+bar_width,count3,bar_width,alpha=opacity,color='g',label='second class')
#type4=plt.bar(index+bar_width+bar_width+bar_width,count4,bar_width,alpha=opacity,color='y',label='same day')
#plt.xlabel('Order Regions')
#plt.ylabel('Number of shipments')
#plt.title('Different Types of shipping methods used in all regions')
#plt.legend()
#plt.xticks(index+bar_width,names,rotation=90)
#plt.tight_layout()
#plt.show()

# Customer Segmentation

NOTE: JUST LIKE THE INTITAL SECTION, THIS SECTION IS ALSO COMMENTED OUT.<p>
Understanding customer needs and targeting specific clusters of customers based on their need is one way for a supply chain company to increase number of customers and also to gain more profits.Since,purchase history of customers is already avaialble in the dataset, it can use RFM analysis for customer segmention. Even though there are so many different methods for customer segmentation,RFM analysis is being used because it utilizes numerical values to show Customer recency,frequency and monetary values and also the output results are easy to interpret.

In [29]:
#Calculating total price for which each order
#data['TotalPrice'] = data['Order Item Quantity'] * data['Order Item Total']# Multiplying item price * Order quantity

In [30]:
#data['order date (DateOrders)'].max() # Calculating when the last order come to check recency

In [31]:
#Present date was set to next day of the last order. i.e,2018-02-01
#present = dt.datetime(2018,2,1)
#data['order date (DateOrders)'] = pd.to_datetime(data['order date (DateOrders)'])

In [32]:
# Grouping all values into new data frame named customer segmentation
#Customer_seg = data.groupby('Order Customer Id').agg({'order date (DateOrders)': lambda x: (present - x.max()).days, 'Order Id': lambda x: len(x), 'TotalPrice': lambda x: x.sum()})
#Changing order dates to int format
#Customer_seg['order date (DateOrders)'] = Customer_seg['order date (DateOrders)'].astype(int)
# Renaming columns as R_Value,F_Value,M_Value
#Customer_seg.rename(columns={'order date (DateOrders)': 'R_Value', 
  #                         'Order Id': 'F_Value', 
 #                        'TotalPrice': 'M_Value'}, inplace=True)
#Customer_seg.head()

In [33]:
#plt.figure(figsize=(12,10)) # Figure size
#plt.subplot(3, 1, 1)
#sns.distplot(Customer_seg['R_Value'])# Plot distribution of R_Value
#plt.subplot(3, 1, 2) 
#sns.distplot(Customer_seg['F_Value'])# Plot distribution of F_Value
#plt.subplot(3, 1, 3)
#sns.distplot(Customer_seg['M_Value'])# Plot distribution of M_Value
#plt.show()

In [34]:
#quantiles = Customer_seg.quantile(q=[0.25,0.5,0.75]) #Dividing RFM data into four quartiles
#quantiles = quantiles.to_dict() 

In [35]:
# R_Score should be minimum so 1st quantile is set as 1.
#def R_Score(a,b,c):
#    if a <= c[b][0.25]:
 #       return 1
 #   elif a <= c[b][0.50]:
 #       return 2
 #   elif a <= c[b][0.75]: 
  #      return 3
   # else:
    #    return 4
# The higher the F_Score,M_Score the better so 1st quantile is set as 4.    
#def FM_Score(x,y,z):
 #   if x <= z[y][0.25]:
  #      return 4
   # elif x <= z[y][0.50]:
    #    return 3
   # elif x <= z[y][0.75]: 
    #    return 2
    #else:
     #   return 1

In [36]:
# New column for R_Score to indicate numerical score between 1 to 4.
#Customer_seg['R_Score'] = Customer_seg['R_Value'].apply(R_Score, args=('R_Value',quantiles))
# New column for F_Score to indicate numerical score between 1 to 4.
#Customer_seg['F_Score'] = Customer_seg['F_Value'].apply(FM_Score, args=('F_Value',quantiles))
# New column for M_Score to indicate numerical score between 1 to 4.
#Customer_seg['M_Score'] = Customer_seg['M_Value'].apply(FM_Score, args=('M_Value',quantiles))
#Customer_seg.head()

The individual scores of R,F,M are known.A column for combined RFM score is created.

In [37]:
#Adding R,F,M Scores to one new column
#Customer_seg['RFM_Score'] = Customer_seg.R_Score.astype(str)+ Customer_seg.F_Score.astype(str) + Customer_seg.M_Score.astype(str)
#Customer_seg.head()

In [38]:
#count=Customer_seg['RFM_Score'].unique()
#print(count)# Printing all Unique values
#len(count)# Total count

In [39]:
# Calculate RFM_Score
#Customer_seg['RFM_Total_Score'] = Customer_seg[['R_Score','F_Score','M_Score']].sum(axis=1)
#Customer_seg['RFM_Total_Score'].unique()

In [40]:
# Define rfm_level function
#def RFM_Total_Score(df):

    #if (df['RFM_Total_Score'] >= 11):# For RFM score with values 11,12
      #  return 'Champions' 
   # elif (df['RFM_Total_Score'] == 10):# For RFM score with value 10
     #   return 'Loyal Customers' 
    #elif (df['RFM_Total_Score'] == 9): # For RFM score with value 9
    #    return 'Recent Customers'
    #elif (df['RFM_Total_Score'] == 8): # For RFM score with value 8
   #      return 'Promising'
   # elif (df['RFM_Total_Score'] == 7): # For RFM score with value 7
    #    return 'Customers Needing Attention'
    #elif (df['RFM_Total_Score'] == 6): # For RFM score with value 6
    #    return 'Cant lose them'
    #elif (df['RFM_Total_Score'] == 5): # For RFM score with value 5
     #   return 'At Risk'
    #else:                               # For RFM score with value less than 5
        
        #return 'Lost'
# Create a new variable RFM_Level
#Customer_seg['Customer_Segmentation'] =Customer_seg.apply(RFM_Total_Score, axis=1)
# Print the header with top 5 rows to the console
#Customer_seg.head()


In [41]:
# Calculate average values for each RFM_Level, and return a size of each segment 
#Customer_seg['Customer_Segmentation'].value_counts().plot.pie(figsize=(10,10),
                                     #   startangle=135, explode=(0,0,0,0.1,0,0,0,0),autopct='%.1f',shadow=True)
#plt.title("Customer Segmentation",size=15)
#plt.ylabel(" ")
#plt.axis('equal') 
#plt.show()

In [42]:
#churned=Customer_seg[(Customer_seg['RFM_Score']=='411')].sort_values('M_Value', ascending=False).head(10)
#churned

In [43]:
#The R_Score should be low and F_Score, M_Score should be as high as possible
#Customer_seg[(Customer_seg['RFM_Score']=='144')|(Customer_seg['RFM_Score']=='143')].sort_values('M_Value', ascending=False).head(10)


# Data Modelling

To measure the performance of different models the machine learning models are trained to detect fraud,late delivery for classification type. And sales, order quantity is predicted for regression type models.

A new dataset is created with the copy of original data for training the data and validation.

In [47]:
train_data=data.copy()

Two new columns are created for orders with suspected fraud and late delivery making them into binary classification, which in turn helps to measure performance of different models better.

In [48]:
train_data['fraud'] = np.where(train_data['Order Status'] == 'SUSPECTED_FRAUD', 1, 0)
train_data['late_delivery']=np.where(train_data['Delivery Status'] == 'Late delivery', 1, 0)

Now to measure machine models accurately all the columns with repeated values are dropped like late_delivery_risk column because, it is known all the products with late delivery risk are delivered late. And Order Status column because, a new column for fraud detection is created there is a chance machine learning model might take values directly from these columns to predict output.

In [49]:
#Dropping columns with repeated values
train_data.drop(['Delivery Status','Late_delivery_risk','Order Status','order_month_year','order date (DateOrders)'], axis=1, inplace=True)

It is important to check the type of variables in the data because machine learning models can only be trained with numerical values.

In [50]:
train_data.dtypes

Type                                 str
Days for shipping (real)           int64
Days for shipment (scheduled)      int64
Benefit per order                float64
Sales per customer               float64
Category Id                        int64
Category Name                        str
Customer City                        str
Customer Country                     str
Customer Id                        int64
Customer Segment                     str
Customer State                       str
Customer Zipcode                 float64
Department Id                      int64
Department Name                      str
Market                               str
Order City                           str
Order Country                        str
Order Customer Id                  int64
Order Id                           int64
Order Item Cardprod Id             int64
Order Item Discount              float64
Order Item Discount Rate         float64
Order Item Id                      int64
Order Item Produ

There are some columns with object type data which cannot be trained in machine learning models so all the object type data is converted to int type using preprocessing label encoder library.

In [51]:
# create the Labelencoder object
le = preprocessing.LabelEncoder()
#convert the categorical columns into numeric
train_data['Customer Country']  = le.fit_transform(train_data['Customer Country'])
train_data['Market']            = le.fit_transform(train_data['Market'])
train_data['Type']              = le.fit_transform(train_data['Type'])
train_data['Product Name']      = le.fit_transform(train_data['Product Name'])
train_data['Customer Segment']  = le.fit_transform(train_data['Customer Segment'])
train_data['Customer State']    = le.fit_transform(train_data['Customer State'])
train_data['Order Region']      = le.fit_transform(train_data['Order Region'])
train_data['Order City']        = le.fit_transform(train_data['Order City'])
train_data['Category Name']     = le.fit_transform(train_data['Category Name'])
train_data['Customer City']     = le.fit_transform(train_data['Customer City'])
train_data['Department Name']   = le.fit_transform(train_data['Department Name'])
train_data['Order State']       = le.fit_transform(train_data['Order State'])
train_data['Shipping Mode']     = le.fit_transform(train_data['Shipping Mode'])
train_data['order_week_day']    = le.fit_transform(train_data['order_week_day'])
train_data['Order Country']     = le.fit_transform(train_data['Order Country'])
train_data['Customer Full Name']= le.fit_transform(train_data['Customer Full Name'])

#display the initial records
train_data.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Category Id,Category Name,Customer City,Customer Country,Customer Id,...,Product Name,Product Price,Shipping Mode,Customer Full Name,order_year,order_month,order_week_day,order_hour,fraud,late_delivery
0,1,3,4,91.250000,314.640015,73,40,66,1,20755,...,78,327.75,3,1875,2018,1,6,22,0,0
1,3,5,4,-249.089996,311.359985,73,40,66,1,19492,...,78,327.75,3,5374,2018,1,2,12,0,1
2,0,4,4,-247.779999,309.720001,73,40,452,0,19491,...,78,327.75,3,4426,2018,1,2,12,0,0
3,1,3,4,22.860001,304.809998,73,40,285,0,19490,...,78,327.75,3,12922,2018,1,2,11,0,0
4,2,2,4,134.210007,298.250000,73,40,66,1,19489,...,78,327.75,3,10632,2018,1,2,11,0,0


Now all the data is transformed into int type. The dataset is split into train data and test data so model can be trained with train data and the performance of model can be evaluated using test data.

## Comparision of Classification Models

In [52]:
#All columns expect fraud
xf=train_data.loc[:, train_data.columns != 'fraud']
#Only fraud column
yf=train_data['fraud']
#Splitting the data into two parts in which 80% data will be used for training the model and 20% for testing
xf_train, xf_test,yf_train,yf_test = train_test_split(xf,yf,test_size = 0.2,random_state = 42)
#All columns expect fraud
xl=train_data.loc[:, train_data.columns != 'late_delivery']
#Only fraud column
yl=train_data['late_delivery']
#Splitting the data into two parts in which 80% data will be used for training the model and 20% for testing
xl_train, xl_test,yl_train,yl_test = train_test_split(xl,yl,test_size = 0.2, random_state = 42)

Since there are so many different variables with different ranges standard scaler is used to standardize total the data so it is internally consistent before training the data with machine learning. Training after splitting also prevents data leakage which would risk overfitting.

In [53]:
sc = StandardScaler()
xf_train=sc.fit_transform(xf_train)
xf_test=sc.transform(xf_test)
xl_train=sc.fit_transform(xl_train)
xl_test=sc.transform(xl_test)

The data is now ready to be used in machine learning models since many different models are compared training every model from begining is complicated so a function is defined to make the process bit easy. The output is in binary classification format so all the models are measured with Accuracy score,recall score and F1 score metrics. 

To measure the performance of different models F1 score is used as the main metric because it is the harmonic mean of precison score and recall score.And all the scores are multiplied with 100 for better understanding

In [54]:
def classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test):
    model_f=model_f.fit(xf_train,yf_train) # Fitting train data for fraud detection
    model_l=model_l.fit(xl_train,yl_train) # Fitting train data for predection of late delivery
    yf_pred=model_f.predict(xf_test)
    yl_pred=model_l.predict(xl_test)  
    accuracy_f=accuracy_score(yf_pred, yf_test) #Accuracy for fraud detection
    accuracy_l=accuracy_score(yl_pred, yl_test) #Accuracy for predection of late delivery
    recall_f=recall_score(yf_pred, yf_test) #Recall score for  fraud detection
    recall_l=recall_score(yl_pred, yl_test)# Recall score for predection of late delivery
    conf_f=confusion_matrix(yf_test, yf_pred)# fraud detection
    conf_l=confusion_matrix(yl_test, yl_pred)#predection of late delivery
    f1_f=f1_score(yf_test, yf_pred)#fraud detection
    f1_l=f1_score(yl_test, yl_pred)#predection of late delivery
    print('Model paramters used are :',model_f)
    print('Accuracy of fraud status is        :', (accuracy_f)*100,'%')
    print('Recall score of fraud status is        :', (recall_f)*100,'%')
    print('Conf Matrix of fraud status is        :\n',  (conf_f))
    print('F1 score of fraud status is        :', (f1_f)*100,'%')
    print('Accuracy of late delivery status is:', (accuracy_l)*100,'%')
    print('Recall score of late delivery status is:', (recall_l)*100,'%')
    print('Conf Matrix of late delivery status is: \n',(conf_l))
    print('F1 score of late delivery status is:', (f1_l)*100,'%')

## CLASSIFIER ASSESMENT <p>
For this assignemnt I chose to asses decision trees, therefor all other classifiers were commented out.

### Logistic classification model

In [55]:
#model_f = LogisticRegression(solver='lbfgs',random_state=0) #the classification model
#model_l = LogisticRegression(solver='lbfgs',random_state=0) #the classification model
#Giving inputs to the defined function
#classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test)

### Gaussian naive bayes model

In [56]:
#model_f = GaussianNB()
#model_l = GaussianNB()
#classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test)

### Support vector machines

In [57]:
#model_f = svm.LinearSVC()
#model_l = svm.LinearSVC()
#classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test)

### K nearest Neighbors classification

In [58]:
#model_f = KNeighborsClassifier(n_neighbors=1)
#model_l = KNeighborsClassifier(n_neighbors=1)
#classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test)

### Linear Discriminant Analysis

In [59]:
#model_f = LinearDiscriminantAnalysis()
#model_l = LinearDiscriminantAnalysis()
#classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test)

### Random forest classification

In [60]:
#model_f = RandomForestClassifier()
#model_l = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=0)
#classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test)

### Extra trees classification

In [61]:
#model_f = ExtraTreesClassifier(n_estimators=100, max_depth=None,random_state=0)
#model_l = ExtraTreesClassifier(n_estimators=100, max_depth=None,random_state=0)
#classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test)

### eXtreme Gradient Boosting Classification

In [62]:
#model_f = xgb.XGBClassifier()
#model_l = xgb.XGBClassifier()
#classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test)

### Decision tree classification

In [66]:
#model_f = tree.DecisionTreeClassifier()
#model_l = tree.DecisionTreeClassifier()
#classifiermodel(model_f,model_l,xf_train, xf_test,yf_train,yf_test,xl_train, xl_test,yl_train,yl_test)

For better understanding and comparision of all the scores a dataframe is created

In [69]:
#Giving column Values
#classification_data = {'Classification Model':  ['Logistic','Gausian Naive bayes','Support Vector Machines','K nearest Neighbour',
   #                             'Linear Discriminant Analysis','Random Forest','Extra trees','eExtreme gradient boosting','Decision tree'],
      #  'Accuracy Score for Fraud Detection':     [97.80,87.84,97.75,97.36,97.88,98.48,98.61,98.93,99.12],
      #  'Recall Score for Fraud Detection':       [59.40,16.23,56.89,41.90,56.57,93.18,98.88,89.89,82.53],
      #  'F1 Score for Fraud Detection':           [31.22,27.92,28.42,35.67,49.20,54.57,58.60,73.22,81.00],
       # 'Accuracy Score for Late Delivery':       [98.84,57.27,98.84,80.82,98.37,98.60,99.17,99.24,99.37],
       # 'Recall Score for Late Delivery':         [97.94,56.20,97.94,83.45,97.68,97.52,98.51,98.65,99.44],
       # 'F1 Score for Late Delivery':             [98.96,71.95,98.96,82.26,98.52,98.74,99.25,99.31,99.42] }
#Creating data frame with Column Names
#classification_comparision = pd.DataFrame (classification_data, columns = ['Classification Model','Accuracy Score for Fraud Detection','Recall Score for Fraud Detection','F1 Score for Fraud Detection',
 #                                                                          'Accuracy Score for Late Delivery','Recall Score for Late Delivery','F1 Score for Late Delivery'])


### Comparision Table for Classification Scores

In [68]:
#classification_comparision #Printing dataframe

In [67]:
#classification_comparision.max() #Checking max values in every column

### Cross validation

In [70]:
#Defining cross validation model
#def cross_validation_model(model_f,model_l,xf,yf,xl,yl):
#    model_f= model_f.fit(xf,yf)
   # model_l = model_l.fit(xl,yl)
   # scores_f = cross_val_score(model_f, xf, yf, cv=6)
    #scores_l = cross_val_score(model_l, xl, yl, cv=6)
    #print('Model used is',model_f)
    #print('Cross validation accuracy of fraud: %0.2f (+/- %0.2f)' % (scores_f.mean(), scores_f.std() * 2))
    #print('Cross validation accuracy of late : %0.2f (+/- %0.2f)' % (scores_l.mean(), scores_l.std() * 2))

In [71]:
#cross_validation_model(model_f,model_l,xf,yf,xl,yl)

Since, the difference between cross validated scores and accuracy scores of the model is very minimal it can be confirmed that the data is neither overfitted or underfitted, Which variable was given more importance in the model is found using feature importance method from sklearn.

### Feature Importance

In [72]:
#important_col=model_f.feature_importances_.argsort()
#feat_imp=pd.DataFrame({'Variables':xf.columns[important_col],'importance':model_f.feature_importances_[important_col]})
#feat_imp=feat_imp.sort_values(by='importance',ascending=False)
#ax = sns.catplot(x='Variables', y = 'importance', data=feat_imp, height=5, aspect=2,  kind="bar")
#plt.xticks(rotation=90)

In [73]:
#important_col=model_l.feature_importances_.argsort()
#feat_imp=pd.DataFrame({'features':xl.columns[important_col],'importance':model_l.feature_importances_[important_col]})
#feat_imp=feat_imp.sort_values(by='importance',ascending=False)
#ax = sns.catplot(x='features', y = 'importance', data=feat_imp, height=5, aspect=2,  kind="bar")
#plt.xticks(rotation=90)

It can be seen that the columns for the days of shipping is given almost 90% importance in decision tree model,it will be interesting to see how well the model can predict when these variables are removed.

So a new model with the copy of train data is created
.

In [74]:
#new_data=train_data.copy()

In [75]:
# Dropping columns in new data set
#new_data=new_data.drop(['Days for shipping (real)','Days for shipment (scheduled)'],axis=1)

In [76]:
#All columns expect fraud
#new_xl=new_data.loc[:, new_data.columns != 'late_delivery']
##Only fraud column
#new_yl=train_data['late_delivery']
#Splitting the data into two parts in which 80% data will be used for training the model and 20% for testing
#new_xl_train, new_xl_test,new_yl_train,new_yl_test = train_test_split(new_xl,new_yl,test_size = 0.2,random_state = 42)

Standardizing data with Standardscaler module:

In [77]:
#new_xl_train=sc.fit_transform(new_xl_train)
#new_xl_test=sc.transform(new_xl_test)

Function for classification model is created to train one model

In [78]:
#def New_classifiermodel(model_c,xc_train, xc_test,yc_train,yc_test):
 #   model_c=model_c.fit(xc_train,yc_train)
  #  yc_pred=model_c.predict(xc_test) 
   # accuracy_c=accuracy_score(yc_pred, yc_test)
    #recall_c=recall_score(yc_pred, yc_test,average='weighted')
    #conf_c=confusion_matrix(yc_test, yc_pred)
    #f1_c=f1_score(yc_test, yc_pred,average='weighted')
    #print('Model paramters used are :',model_c)
    #print('Accuracy         :', (accuracy_c)*100,'%')
    #print('Recall score        :', (recall_c)*100,'%')
    #print('Conf Matrix        : \n',(conf_c))
    #print('F1 score       :', (f1_c)*100,'%')

### Decision Tree classification

In [79]:
#new_model_l=tree.DecisionTreeClassifier()
#New_classifiermodel(new_model_l,new_xl_train, new_xl_test,new_yl_train,new_yl_test)

Even when shipping days variables were removed the F1 score and the accuracy of the new model is nearly 84% which is still pretty good. Which variables are given more importance this time?

In [80]:
#important_col=new_model_l.feature_importances_.argsort()
#feat_imp=pd.DataFrame({'features':new_xl.columns[important_col],'importance':new_model_l.feature_importances_[important_col]})
#feat_imp=feat_imp.sort_values(by='importance',ascending=False)
#ax = sns.catplot(x='features', y = 'importance', data=feat_imp, height=5, aspect=2,  kind="bar")
#plt.xticks(rotation=90)

In [81]:
#All columns expect order country
#xc=train_data.loc[:, train_data.columns != 'Order Country']
#Order column country
#yc=train_data['Order Country']
#Splitting 20% of dataset as test data 
#xc_train, xc_test,yc_train,yc_test = train_test_split(xc,yc,test_size = 0.2, random_state = 42)

In [82]:
#xc_train=sc.fit_transform(xc_train)
#xc_test=sc.transform(xc_test)

### Decision Tree Multi Classification Model

In [83]:
#model_c=tree.DecisionTreeClassifier()
#New_classifiermodel(model_c,xc_train, xc_test,yc_train,yc_test)

### Neural Network Model for Classification

In [88]:
keras.layers.BatchNormalization()
classifier = Sequential()
#First Hidden Layer
classifier.add(Dense(1024, activation='relu',kernel_initializer='random_normal', input_dim=43)) #Since we have 44 columns
#Third Hidden Layer
classifier.add(Dense(512, activation='relu',kernel_initializer='random_normal'))
#Fourth Hidden Layer
classifier.add(Dense(256, activation='relu',kernel_initializer='random_normal'))
#Fifth Hidden Layer
classifier.add(Dense(128, activation='relu',kernel_initializer='random_normal'))
#Sixth Hidden Layer
classifier.add(Dense(64, activation='relu',kernel_initializer='random_normal'))
#Seventh Hidden Layer
classifier.add(Dense(32, activation='relu',kernel_initializer='random_normal'))
#Eight Hidden Layer
classifier.add(Dense(16, activation='relu',kernel_initializer='random_normal'))
#Ninth Hidden Layer
classifier.add(Dense(8, activation='relu',kernel_initializer='random_normal'))
#Tenth Hidden Layer
classifier.add(Dense(4, activation='relu',kernel_initializer='random_normal'))
#Eleventh Hidden Layer
classifier.add(Dense(2, activation='relu',kernel_initializer='random_normal'))
#Output Layer
classifier.add(Dense(1, activation='sigmoid',kernel_initializer='random_normal'))

Since output data is binary classification the binary_crossentropy is used to measure loss and accuracy is used as metric to train the model because F1 score is not available in Keras.

In [89]:
classifier.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

The model is trained with batch size of 512 and 10 epochs.

In [90]:
#Fitting the data to the training dataset
classifier.fit(xf_train,yf_train, batch_size=512, epochs=10)

Epoch 1/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 11s 24ms/step - accuracy: 0.9778 - loss: 0.5840
Epoch 2/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 7s 26ms/step - accuracy: 0.9778 - loss: 0.3164
Epoch 3/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 9s 31ms/step - accuracy: 0.9778 - loss: 0.1700
Epoch 4/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 7s 24ms/step - accuracy: 0.9778 - loss: 0.1245
Epoch 5/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 8s 29ms/step - accuracy: 0.9778 - loss: 0.1116
Epoch 6/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 8s 27ms/step - accuracy: 0.9778 - loss: 0.1079
Epoch 7/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 9s 31ms/step - accuracy: 0.9778 - loss: 0.1069
Epoch 8/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 10s 34ms/step - accuracy: 0.9778 - loss: 0.1067
Epoch 9/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 8s 27ms/step - accuracy: 0.9778 - loss: 0.1067
Epoch 10/10
283/283 ━━━━━━━━━━━━━━━━━━━━ 9s 33ms/step - accuracy: 0.9778 - loss: 0.1066


It can be seen that the neural network model is performing better with every epoch even tough accuracy remained same the loss is decreasing with every epoch.Since every cell results is saved in jupyter notebook for the next iteration the model is trained with 30 more epochs and the results are displayed.<p>

The cell below  tests the model's ability to identify fraud by processing the data in 512-row batches across 30 training cycles. Accuracy improves as the **binary crossentropy loss** measures prediction errors, which the **adam optimizer** then uses to fine-tune internal weights.



In [91]:
classifier.fit(xf_train,yf_train, batch_size=512, epochs=30)

Epoch 1/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 7s 24ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 2/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 7s 25ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 3/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 9s 30ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 4/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 9s 30ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 5/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 8s 30ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 6/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 9s 33ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 7/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 9s 31ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 8/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 9s 30ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 9/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 10s 36ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 10/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 9s 31ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 11/30
283/283 ━━━━━━━━━━━━━━━━━━━━ 10s 37ms/step - accuracy: 0.9778 - loss: 0.1066
Epoch 12/30
283/283 ━━━━━━━━━━━━━━━━━━━

The model is evaluated with test data set

In [92]:
train_evaluate=classifier.evaluate(xf_train, yf_train)
test_evaluate=classifier.evaluate(xf_test, yf_test)
print('accuracy for Train set is',train_evaluate)
print('accuracy for Test set is',test_evaluate)# evaluation of model.
yf_pred1=classifier.predict(xf_test,batch_size=512,verbose=1)
yf_pred=np.argmax(yf_pred1,axis=1)
print(f1_score(yf_test,yf_pred,average="weighted"))

4513/4513 ━━━━━━━━━━━━━━━━━━━━ 14s 3ms/step - accuracy: 0.9778 - loss: 0.1066
1129/1129 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.9765 - loss: 0.1115
accuracy for Train set is [0.10664932429790497, 0.977758526802063]
accuracy for Test set is [0.1115327924489975, 0.976456880569458]
71/71 ━━━━━━━━━━━━━━━━━━━━ 1s 10ms/step
0.9648255733921929


# Conclusion

The Neural Network achieved a Test Accuracy of 97.65% and a weighted F1-score of 96.48%. The minimal gap between training and testing metrics (less than 0.2%) confirms the model's ability to accurately classify unseen supply chain transactions without overfitting